compiler/semstmts.nim

```diff
@@ -101,6 +103,9 @@ proc semWhile(c: PContext, n: PNode; flags: TExprFlags): PNode =
     result.typ = c.enforceVoidContext
   elif efInTypeof in flags:
     result.typ = n[1].typ
+  elif implicitlyDiscardable(n[1]):
+    result[1].typ = c.enforceVoidContext
```

Comment:
Why would result[1].typ be the void type? `discard producesVoid()` is not allowed.

Reply:
Since it is an implicit `discardable` call, there is no discard. So I simply set the type of the whole body to void type.

```nim
  while true:
    hello()
 ```

`result[1].typ` corresponds to the type of the while statement body, namely the type of `discardable` call. Since an implicit `discardable` call won't be rewritten to a discard statement, we need to set the type of the body to void type or plus turn the `discardable` call into a discard statement.